src/System.Linq.Expressions/src/System/Dynamic/ExpandoObject.cs

```diff
@@ -246,15 +247,14 @@ internal bool IsDeletedMember(int index)
         private ExpandoData PromoteClassCore(ExpandoClass oldClass, ExpandoClass newClass)
         {
             Debug.Assert(oldClass != newClass);
+            ContractUtils.AssertLockHeld(LockObject);
 
-            lock (LockObject)
+            if (_data.Class == oldClass)
```

Comment:
I'm not sure the assert is valid in that case, as it could lose in the same race condition. Still, it's not touched by this PR and it won't affect release code, so I'll look at that possibility later.